Add live mode

cvn opened the pull request:
Hi Igor,

This is a pretty big PR (again), but it started with simple goals:

1. make the scope work with mono files
2. add a "live" mode, where the scope data comes directly from the playing audio

Working on these led to other changes. You can read the commit messages for the details, but it's probably more fun to run the demo first http://cvn.github.io/woscope/. You'll see new options at the bottom of the page.

Please feel free to ask questions or offer critique.

Chad

---

Notes:

* Make sure to run `npm update` after you pull, I updated some of the dependencies.

* There are two parallel implementations of the live mode, one that uses `audioCtx.createAnalyser` and one that uses `audioCtx.createScriptProcessor`. The analyser path is more performant and generally better, but it's not fully supported in Safari. The best I could do with an analyser in Safari looks like this. Notice the jaggies.

    <p align="center"><img src="https://cloud.githubusercontent.com/assets/774864/22493997/8c5da616-e7e8-11e6-9838-8bda803cfdef.png" width="300" /></p>

    So Safari uses the scriptProcessor path instead, and looks great. But if you'd rather not have the [extra code](/m1el/woscope/pull/6/commits/3bdb853) necessary for this, I have another branch that excludes it: [cvn/woscope/analysers](/cvn/woscope/commits/analysers).

* The dist files now have a bunch of variables prepended by _'s. This is mostly due to changes in `babel-preset-es2015`, particularly the `babel-plugin-transform-es2015-block-scoping` submodule. I looked for a way to get back the old behavior, but it involved either adding more babel dependencies or reorganizing code. I didn't think it was worth it, at least for this PR.

cvn committed:
Add support for Microsoft Edge

cvn committed:
Safari createMediaElementSource volume workaround

cvn committed:
add error callback, refactor nogl error

cvn committed:
don't throw web audio api error on load

- since errors stop execution in a js file, if woscope was concatenated with other files for deployment, it could break an app. this is fixed now.
- error on woscope invocation instead
- send web audio API errors to error callback

cvn committed:
make bloom accessible from config, rename doBloom to bloom

cvn committed:
add options UI to demo (swap, invert, bloom)

cvn committed:
ctx.swap parameter now works during playback

- refactor loadWaveAtPosition to use subarray

cvn committed:
fix exception with mono files by duplicating ch0 to ch1

- "Uncaught DOMException: Failed to execute 'getChannelData' on 'AudioBuffer': channel index (1) exceeds number of channels (1)"
- matches behavior of createMediaElementSource with mono files
- add audioData.sourceChannels for communicating that source is mono, even though the data becomes stereo

cvn committed:
make foreground color configurable

- fsLine.glsl: remove redundant "precision highp float;"

cvn committed:
add sweep mode, for more traditional oscilloscope display

- add vbo2, color2 for 2nd channel in sweep mode

cvn committed:
refactor drawLine

cvn committed:
add live option to demo page

- feature to be implemented in following commit
- allow volume to be 100% because it affects display size for live mode

cvn committed:
add destroy function, call it to clean up

cvn committed:
add live mode, visualize in real time without preloading file into buffer

cvn committed:
update dependencies

- update budo
  - fix page load flakiness from external devices
  - fix "Possible EventEmitter memory leak detected" console error
- remove bannerify
  - was generating banner incorrectly with new versions of node
  - https://github.com/zemirco/bannerify/issues/2
- add browserify-banner
  - replaces bannerify, works correctly for this project

cvn committed:
send audio file errors to error callback

cvn committed:
woscope returns context object

cvn committed:
make background color configurable

cvn committed:
make live mode compatible with Safari

- use createScriptProcessor instead of createAnalyser because Safari does not support getFloatTimeDomainData, and emulating with getByteTimeDomainData gives low resolution results

m1el commented:
Thank you for the PR, it seems that you put more hours in this demo than I did.

From the first glance, this seems good and works nice :)

Here's a few thoughts on the changes:

There's a problem with configuring background color: I'm using a blending function that "collects" or "adds" the color instead of using proper alpha blending:

https://github.com/cvn/woscope/blob/1b4e848a471385474762ac613da7bbb3bbee9b9d/index.js#L642

```javascript
    gl.blendFunc(gl.SRC_ALPHA, gl.ONE);
```

If the background color is anything but black, this will produce weird results.  E.g. you'll see no picture if the background is white.

Changing the foreground color will produce unexpected colors because of over-saturation.  I carefully picked the foreground color so that it over-saturates to white at a nice rate.  With GLES3 it would be possible to collect the intensity to a monochrome float texture, and this is a little bit of hack that can produce similar results (although, with invalid gamma).  I don't know a nice way to produce similar results using a single draw call in GLES2.  I think it's possible to change the color in post-processing, probably even correct the gamma.

I'll try to inspect these changes in detail soon.

- Igor

cvn commented:
I understand what you're saying about the background/foreground colors. I made them configurable because I wanted 1) a transparent background and 2) different colors for each channel in sweep mode. 

I don't mind the limitations of additive blending. I think the visual choices you made are part of the reason why I like woscope so much, so I would be wary of any big changes to the compositing.

Additionally, thanks to CSS filters, it's already straightforward to achieve a dark color on a light background by applying `filter: invert(1);` to the canvas. The end result may not be the world's most intuitive API for choosing colors, but it feels more like you're working with a real oscilloscope, at least to me.

m1el commented:
Please make the following changes:

- With live audio, you're using `analyser.fftSize` as buffer size, but it's not set anywhere.  While the default value for `fftSize` is specified in the spec, I'd like it to be set in the code.  We could probably use a power of two which is closest to `nSamples`?
- When drawing a mono file in sweep mode, should we draw two sweep lines?
- Add your name to `LICENSE` file if you want to.

Great job on adding a live processing mode, it's requires less RAM and loads much faster.

cvn committed:
set analyser fftSize explicitly

- using default value of 2048. it looks as good as 4096 (nSamples) and performs better
- also make analyser channel connection more explicit

cvn committed:
update license

cvn committed:
Release 0.2.0

cvn commented:
OK. I pushed some new commits.

* I set fftSize to the default size of 2048. I tried 4096 (nSamples), but it performed worse on Edge running under VMWare Fusion (60fps vs ~20fps). I didn't see any quality difference between 2048 and 4096 in Chrome, where both sizes ran smoothly. I tried 1024 too, but the scope looks patchy in xy mode and in sweep mode it's partially blank.
* In live mode, it's not easy to get the number of audio channels from the source<sup>1</sup>, so I always assume stereo audio and draw 2 lines. I considered adding a mono mode, for manual control, but I decided that setting `color2: [0,0,0,0]` is good enough and keeps things simpler<sup>2</sup>.
* I'm proud to add my name to the license, thank you.

---
<sup>1</sup> `createMediaElementSource` nodes don't report the number of source channels, only the number of output channels, WebAudio/web-audio-api#286. People suggest decoding the file to get the info, which implies there's no easier way, http://stackoverflow.com/q/32835149.

<sup>2</sup> Even though I didn't add a mono mode, there's at least one good argument for having one. There are cross-compatibility issues with mono files and `createChannelSplitter`, one of the nodes in the analyser path. Chrome outputs the mono data over both channels, Safari/Firefox/Edge only output data on ch0, and ch1 is silent. This is reflected in the final output as well, so mono files only play from the left speaker in Safari/FF/Edge. This could be worked around with a mono mode. (Safari is not affected in practice because it uses the scriptProcessor path.) I may file bugs to try to convince browser makers to align their behavior, but I don't know how likely that is, http://stackoverflow.com/questions/14915951/web-audio-no-sound-in-right-channel#comment20967458_14934906.
